### PR TITLE
Implement intermediate code compiler

### PR DESCRIPTION
**Pull request by m3m0r7**
# TODO
- [x] Add validation into some methods.
- [x] Add Access Flag generator (Class/Method)
- [x] Add argument generator
- [x] Update README.md
- [x] Update Version
- [x] Add test

### CODE REVIEW

**Comment by memory-agape** [comment]
A Test code:

```
<?php
require __DIR__ . '/../vendor/autoload.php';

use PHPJava\Compiler\Builder\Attributes;
use PHPJava\Compiler\Builder\ConstantPool;
use PHPJava\Compiler\Builder\Methods;
use PHPJava\Compiler\Builder\Structures\Attribute\Attribute;
use PHPJava\Compiler\Builder\Structures\ClassFileStructure;
use PHPJava\Compiler\Builder\Structures\Info\ClassInfo;
use PHPJava\Compiler\Builder\Structures\Info\MethodInfo;
use PHPJava\Compiler\Builder\Structures\Info\MethodrefInfo;
use PHPJava\Compiler\Builder\Structures\Info\FieldrefInfo;
use PHPJava\Compiler\Builder\Structures\Info\NameAndTypeInfo;
use PHPJava\Compiler\Builder\Structures\Info\StringInfo;
use PHPJava\Compiler\Builder\Structures\Info\Utf8Info;
use PHPJava\Compiler\Compiler;
use PHPJava\Kernel\Maps\MethodAccessFlag;

$compiler = new Compiler(
    (new ClassFileStructure())
        ->setMinorVersion(0)
        ->setMajorVersion(52)
        ->setConstantPool(
            (new ConstantPool())
                ->append(new MethodrefInfo(6, 15))
                ->append(new FieldrefInfo(16, 17))
                ->append(new StringInfo(18))
                ->append(new MethodrefInfo(19, 20))
                ->append(new ClassInfo(21))
                ->append(new ClassInfo(22))
                ->append(new Utf8Info('<init>'), $initNameEntryNumber)
                ->append(new Utf8Info('()V'), $initDescriptorEntryNumber)
                ->append(new Utf8Info('Code'), $codeAttributeEntryNumber)
                ->append(new Utf8Info('LineNumberTable'))
                ->append(new Utf8Info('main'), $mainNameEntryNumber)
                ->append(new Utf8Info('([Ljava/lang/String;)V'), $mainDescriptorEntryNumber)
                ->append(new Utf8Info('SourceFile'), $sourcefileEntryNumber)
                ->append(new Utf8Info('Test.java'), $sourcefileNameEntryNumber)
                ->append(new NameAndTypeInfo(7, 8))
                ->append(new ClassInfo(23))
                ->append(new NameAndTypeInfo(24, 25))
                ->append(new Utf8Info('Hello World!'))
                ->append(new ClassInfo(26))
                ->append(new NameAndTypeInfo(27, 28))
                ->append(new Utf8Info('Test'))
                ->append(new Utf8Info('java/lang/Object'))
                ->append(new Utf8Info('java/lang/System'))
                ->append(new Utf8Info('out'))
                ->append(new Utf8Info('Ljava/io/PrintStream;'))
                ->append(new Utf8Info('java/io/PrintStream'))
                ->append(new Utf8Info('println'))
                ->append(new Utf8Info('(Ljava/lang/String;)V'))
                ->toArray()
        )
        ->setMethods(
            (new Methods())
                ->append(
                    (new MethodInfo(
                        MethodAccessFlag::ACC_PUBLIC,
                        $initNameEntryNumber,
                        $initDescriptorEntryNumber
                    ))
                        ->setAttributes(
                            (new Attributes())
                                ->append(
                                    (new PHPJava\Compiler\Builder\Structures\Attribute\Attribute($codeAttributeEntryNumber))
                                        ->setValue("\x2A\xB7\x00\x01\xB1")
                                )
                                ->toArray()
                        )
                )
                ->append(
                    (new MethodInfo(
                        MethodAccessFlag::ACC_PUBLIC | MethodAccessFlag::ACC_STATIC,
                        $mainNameEntryNumber,
                        $mainDescriptorEntryNumber
                    ))
                        ->setAttributes(
                            (new Attributes())
                                ->append(
                                    (new PHPJava\Compiler\Builder\Structures\Attribute\Attribute($codeAttributeEntryNumber))
                                        ->setValue("\xB2\x00\x02\x12\x00\x03\xB6\x00\x04\xB1")
                                )
                                ->toArray()
                        )
                )
                ->toArray()
        )
        ->setAttributes(
            (new Attributes())
                ->append(
                    (new PHPJava\Compiler\Builder\Structures\Attribute\Attribute($sourcefileEntryNumber))
                        ->setValue(pack('n', $sourcefileNameEntryNumber))
                )
                ->toArray()
        )
);

$compiler->make(fopen('Test.class', 'w+'));

```

**Comment by m3m0r7** [comment]
This is a test code. This is generated intermediate code successfully and possible to run the code with java command.

- 2019/09/14:
  - Added `Finder` class which can find targeted entry.
  - TODO: Refactoring method parameters, Same object internalize.

```php
<?php
require __DIR__ . '/../../vendor/autoload.php';

use PHPJava\Compiler\Builder\Attributes\Architects\Operation;
use PHPJava\Compiler\Builder\Attributes\Code;
use PHPJava\Compiler\Builder\Collection\Attributes;
use PHPJava\Compiler\Builder\Collection\ConstantPool;
use PHPJava\Compiler\Builder\Collection\Methods;
use PHPJava\Compiler\Builder\Attribute;
use PHPJava\Compiler\Builder\Finder\ConstantPoolFinder;
use PHPJava\Compiler\Builder\Signatures\Descriptor;
use PHPJava\Compiler\Builder\Structures\ClassFileStructure;
use PHPJava\Compiler\Builder\Structures\Info\ClassInfo;
use PHPJava\Compiler\Builder\Method;
use PHPJava\Compiler\Builder\Structures\Info\MethodrefInfo;
use PHPJava\Compiler\Builder\Structures\Info\FieldrefInfo;
use PHPJava\Compiler\Builder\Structures\Info\NameAndTypeInfo;
use PHPJava\Compiler\Builder\Structures\Info\StringInfo;
use PHPJava\Compiler\Builder\Structures\Info\Utf8Info;
use PHPJava\Compiler\Builder\Types\Uint16;
use PHPJava\Compiler\Builder\Types\Uint8;
use PHPJava\Compiler\Compiler;
use PHPJava\Kernel\Maps\ClassAccessFlag;
use PHPJava\Kernel\Maps\MethodAccessFlag;
use PHPJava\Kernel\Maps\OpCode;
use PHPJava\Kernel\Resolvers\SDKVersionResolver;
use PHPJava\Kernel\Types\_Void;
use PHPJava\Packages\java\io\PrintStream;
use PHPJava\Packages\java\lang\_String;

$constantPool = new ConstantPool();
$finder = new ConstantPoolFinder($constantPool);

[$majorVersion, $minorVersion] = SDKVersionResolver::resolveByVersion(8);

$compiler = new Compiler(
    (new ClassFileStructure())
        ->setMinorVersion($minorVersion)
        ->setMajorVersion($majorVersion)
        ->setConstantPool(
            $constantPool
                ->add(new Utf8Info('Hello PHPJava Compiler!'))
                ->add(new Utf8Info('Test'))
                ->add(new Utf8Info('java/lang/Object'))
                ->add(new Utf8Info('java/lang/System'))
                ->add(new Utf8Info('out'))
                ->add(new Utf8Info('Ljava/io/PrintStream;'))
                ->add(new Utf8Info('java/io/PrintStream'))
                ->add(new Utf8Info('println'))
                ->add(new Utf8Info('Code'))
                ->add(new Utf8Info('main'))
                ->add(
                    new MethodrefInfo(
                        $finder->find(ClassInfo::class, 'java/lang/Object'),
                        $finder->find(NameAndTypeInfo::class, '<init>', '()V')
                    )
                )
                ->add(
                    new FieldrefInfo(
                        $finder->find(ClassInfo::class, 'java/lang/System'),
                        $finder->find(
                            NameAndTypeInfo::class,
                            'out',
                            (new Descriptor())
                                ->addArgument(PrintStream::class)
                                ->make()
                        )
                    )
                )
                ->add(
                    new StringInfo(
                        $finder->find(Utf8Info::class, 'Hello PHPJava Compiler!')
                    )
                )
                ->add(
                    new MethodrefInfo(
                        $finder->find(ClassInfo::class, 'java/io/PrintStream'),
                        $finder->find(
                            NameAndTypeInfo::class,
                            'println',
                            (new Descriptor())
                                ->addArgument(_String::class)
                                ->setReturn(_Void::class)
                                ->make()
                        )
                    )
                )
                ->add(
                    new ClassInfo(
                        $finder->find(Utf8Info::class, 'Test')
                    )
                )
                ->add(
                    new ClassInfo(
                        $finder->find(Utf8Info::class, 'java/lang/Object')
                    )
                )
                ->add(new Utf8Info('<init>'))
                ->add(new Utf8Info(
                    (new Descriptor())
                        ->setReturn(_Void::class)
                        ->make()
                ))
                ->add(new Utf8Info(
                    (new Descriptor())
                        ->addArgument(_String::class, 1)
                        ->setReturn(_Void::class)
                        ->make()
                ))
                ->add(
                    new NameAndTypeInfo(
                        $finder->find(Utf8Info::class, '<init>'),
                        $finder->find(
                            Utf8Info::class,
                            (new Descriptor())
                                ->setReturn(_Void::class)
                                ->make()
                        )
                    )
                )
                ->add(
                    new ClassInfo(
                        $finder->find(Utf8Info::class, 'java/lang/System')
                    )
                )
                ->add(
                    new NameAndTypeInfo(
                        $finder->find(Utf8Info::class, 'out'),
                        $finder->find(
                            Utf8Info::class,
                            (new Descriptor())
                                ->addArgument(PrintStream::class)
                                ->make()
                        )
                    )
                )
                ->add(
                    new ClassInfo(
                        $finder->find(Utf8Info::class, 'java/io/PrintStream')
                    )
                )
                ->add(
                    new NameAndTypeInfo(
                        $finder->find(Utf8Info::class, 'println'),
                        $finder->find(
                            Utf8Info::class,
                            (new Descriptor())
                                ->addArgument(_String::class)
                                ->setReturn(_Void::class)
                                ->make()
                        )
                    )
                )
                ->add(new Utf8Info(
                    (new Descriptor())
                        ->addArgument(_String::class)
                        ->setReturn(_Void::class)
                        ->make()
                ))
                ->toArray()
        )
        ->setAccessFlags(
            (new \PHPJava\Compiler\Builder\Signatures\ClassAccessFlag())
                ->enableSuper()
                ->make()
        )
        ->setThisClass($finder->find(ClassInfo::class, 'Test'))
        ->setSuperClass($finder->find(ClassInfo::class, 'java/lang/Object'))
        ->setMethods(
            (new Methods())
                ->add(
                    (new Method(
                        (new \PHPJava\Compiler\Builder\Signatures\MethodAccessFlag())
                            ->enablePublic()
                            ->make(),
                        $finder->find(
                            Utf8Info::class,
                            '<init>'
                        ),
                        $finder->find(
                            Utf8Info::class,
                            (new Descriptor())
                                ->setReturn(_Void::class)
                                ->make()
                        )
                    ))
                        ->setAttributes(
                            (new Attributes())
                                ->add(
                                    (new Code($finder->find(Utf8Info::class, 'Code')))
                                        ->setMaxStacks(1)
                                        ->setMaxLocals(1)
                                        ->setCode(
                                            (new Operation())
                                                ->add(OpCode::_aload_0)
                                                ->add(
                                                    OpCode::_invokespecial,
                                                    [
                                                        [
                                                            Uint16::class,
                                                            $finder->find(
                                                                MethodrefInfo::class,
                                                                'java/lang/Object',
                                                                '<init>',
                                                                (new Descriptor())
                                                                    ->setReturn(_Void::class)
                                                                    ->make()
                                                            ),
                                                        ],
                                                    ]
                                                )
                                                ->add(OpCode::_return)
                                        )
                                )
                                ->toArray()
                        )
                )
                ->add(
                    (new Method(
                        (new \PHPJava\Compiler\Builder\Signatures\MethodAccessFlag())
                            ->enablePublic()
                            ->enableStatic()
                            ->make(),
                        $finder->find(Utf8Info::class, 'main'),
                        $finder->find(
                            Utf8Info::class,
                            (new Descriptor())
                                ->addArgument(_String::class, 1)
                                ->setReturn(_Void::class)
                                ->make()
                        )
                    ))
                        ->setAttributes(
                            (new Attributes())
                                ->add(
                                    (new Code($finder->find(Utf8Info::class, 'Code')))
                                        ->setMaxStacks(2)
                                        ->setMaxLocals(1)
                                        ->setCode(
                                            (new Operation())
                                                ->add(
                                                    OpCode::_getstatic,
                                                    [
                                                        [
                                                            Uint16::class,
                                                            $finder->find(
                                                                FieldrefInfo::class,
                                                                'java/lang/System',
                                                                'out',
                                                                (new Descriptor())
                                                                    ->addArgument(PrintStream::class)
                                                                    ->make()
                                                            ),
                                                        ],
                                                    ]
                                                )
                                                ->add(
                                                    OpCode::_ldc,
                                                    [
                                                        [
                                                            Uint8::class,
                                                            $finder->find(
                                                                StringInfo::class,
                                                                'Hello PHPJava Compiler!'
                                                            ),
                                                        ],
                                                    ]
                                                )
                                                ->add(
                                                    OpCode::_invokevirtual,
                                                    [
                                                        [
                                                            Uint16::class,
                                                            $finder->find(
                                                                MethodrefInfo::class,
                                                                'java/io/PrintStream',
                                                                'println',
                                                                (new Descriptor())
                                                                    ->addArgument(_String::class)
                                                                    ->setReturn(_Void::class)
                                                                    ->make()
                                                            ),
                                                        ],
                                                    ]
                                                )
                                                ->add(OpCode::_return)
                                        )
                                )
                                ->toArray()
                        )
                )
                ->toArray()
        )
);

$compiler->compile(fopen('Test.class', 'w+'));
```

**Comment by chitoku-k** [comment]
For the documentation reated fixes, please refer to #229

**Comment by m3m0r7** [comment]
@chitoku-k 
Thanks your reviewing 😭 🙏  You can merge the child branch when you finished rewriting the docs.

File: src/Compiler/Builder/Attribute.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        return $result
+            ->getEntryIndex();
+    }
+
+    public function getValue(): string
+    {
+        return $this->value;
+    }
+
+    public function setValue(string $value): EntryInterface
```
Comment (chitoku-k) [inline src/Compiler/Builder/Attribute.php]:
```suggestion
    public function setValue(string $value): self
```

File: src/Compiler/Builder/Attribute.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        return $this->hasAttribute;
+    }
+
+    public function getAttributes(): array
+    {
+        return $this->attributes;
+    }
+
+    public function setAttributes(array $attributes): EntryInterface
```
Comment (chitoku-k) [inline src/Compiler/Builder/Attribute.php]:
```suggestion
    public function setAttributes(array $attributes): self
```

File: src/Compiler/Builder/Attributes/Architects/Operation.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                    case Uint16::class:
+                        $writer->writeUnsignedShort($value);
+                        break;
+                    case Uint32::class:
+                        $writer->writeUnsignedInt($value);
+                        break;
+                    case Uint64::class:
+                        $writer->writeUnsignedLong($value);
+                        break;
+                }
```
Comment (chitoku-k) [inline src/Compiler/Builder/Attributes/Architects/Operation.php:51]: How about handling unsupported `$type` here?

File: src/Compiler/Builder/Collection/AbstractEntryCollection.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    public function offsetGet($offset)
+    {
+        return $this->entries[$offset];
+    }
+
+    public function offsetSet($offset, $value)
+    {
+        if (!($value instanceof EntryInterface)) {
+            throw new NotAllowedDeleteException('The value is not allowed type.');
```
Comment (chitoku-k) [inline src/Compiler/Builder/Collection/AbstractEntryCollection.php]:
```suggestion
            throw new NotAllowedDeleteException('The value is not allowed.');
```

File: src/Compiler/Builder/Finder/Result/ConstantPoolFinderResult.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                    throw new FinderException('The finder type is not implemented yet. (Type: ' . $this->type . ')');
+            }
+
+            if ($this->isValidText($texts)) {
+                // Returns found index.
+                return $this->resultCaches[$key] = new EntryMap($index, $entry);
+            }
+        }
+
+        throw new FinderException('Not found an entry. (arguments: ' . implode(', ', $this->arguments) . ')');
```
Comment (chitoku-k) [inline src/Compiler/Builder/Finder/Result/ConstantPoolFinderResult.php]:
```suggestion
        throw new FinderException('The entry is not found. (arguments: ' . implode(', ', $this->arguments) . ')');
```

File: src/Compiler/Builder/Maps/EntryMap.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?php
+namespace PHPJava\Compiler\Builder\Maps;
+
+use PHPJava\Compiler\Builder\Structures\Info\AbstractInfo;
+
+class EntryMap implements MapInterface
+{
+    protected $entryIndex;
```
Comment (chitoku-k) [inline src/Compiler/Builder/Maps/EntryMap.php:11]:
```suggestion
    /**
     * @var int
     */
    protected $entryIndex;
```

File: src/Compiler/Builder/Maps/EntryMap.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    protected $entry;
```
Comment (chitoku-k) [inline src/Compiler/Builder/Maps/EntryMap.php:16]:
```suggestion
    /**
     * AbstractInfo
     */
    protected $entry;
```

File: src/Compiler/Builder/Method.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        return $result
+            ->getEntryIndex();
+    }
+
+    public function getAttributes(): array
+    {
+        return $this->attributes;
+    }
+
+    public function setAttributes(array $attributes): EntryInterface
```
Comment (chitoku-k) [inline src/Compiler/Builder/Method.php]:
```suggestion
    public function setAttributes(array $attributes): self
```

File: src/Compiler/Builder/Segments/ConstantPool.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                    case Uint32::class:
+                        $this->binaryWriter->writeUnsignedInt($structureEntry);
+                        break;
+                    case Uint64::class:
+                        $this->binaryWriter->writeUnsignedLong($structureEntry);
+                        break;
+                    case Bytes::class:
+                        $this->binaryWriter->write($structureEntry);
+                        break;
+                }
```
Comment (chitoku-k) [inline src/Compiler/Builder/Segments/ConstantPool.php:42]: Unsupported `$type`?

File: src/Compiler/Builder/Attributes/Architects/Operation.php (hunk 3 above)
Comment (m3m0r7) [reply src/Compiler/Builder/Attributes/Architects/Operation.php:51]: Thanks! I'll fix it

File: src/Compiler/Builder/Maps/EntryMap.php (hunk 7 above)
Comment (m3m0r7) [reply src/Compiler/Builder/Maps/EntryMap.php:16]: Did you forget to add `@var` before AbstractInfo?
I'll fix it.

Comment (chitoku-k) [reply src/Compiler/Builder/Maps/EntryMap.php:16]: Yes, I've totally forgotten!

Comment (m3m0r7) [reply src/Compiler/Builder/Maps/EntryMap.php:16]: Okay! 😄

**Review by chitoku-k — APPROVED** [review]

**Pull request merged**